test.py

```diff
@@ -0,0 +1,226 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import unittest
+from inspect import cleandoc
+from expr_checker import *
+
+class Dangerous:
+    pass
+
+class Good:
+    a = "hi"
+    evil = Dangerous()
+
+    _secret_stuff = "1+1=3"
+
+    def tell_me_hi(self):
+        return self.a
+    
+    def __enter__(self):
+        return Good()
+
+    def __exit__(self, exec_type, exec_value, traceback):
+        pass
+
+
+class ReadOnlyObject:
+    def __init__(self, x):
+        self.set_x_value(x)
+
+    def set_x_value(self, x):
+        self.x = x
+
+
+_WHITELIST = {str, int, type(None), Good, type(range(0))}
+_ALLOWED_ATTR = {"a", "x", "tell_me_hi", "set_x_value"}
+
+
+class TestFuncChecker(unittest.TestCase):
+    def test_function_call(self):
+        def abc(a, b, c):
+            return f"A: {a}, B: {b}, C: {c}"
+
+        self.assertEqual(
+            eval(expr_checker("abc('aaa', 'bbb', 'ccc')",
+                 _WHITELIST, _ALLOWED_ATTR, {})),
+            "A: aaa, B: bbb, C: ccc"
+        )
+
+        with self.assertRaises(ValueError):
+            eval(expr_checker("abc(a='aaa', c='ccc', b=Dangerous())",
+                 _WHITELIST, _ALLOWED_ATTR, {}))
+
+    def test_comp_expr(self):
+        self.assertEqual(
+            eval(expr_checker(
+                "[(lambda x: x**2)(n) for n in range(1, 11)]", _WHITELIST, _ALLOWED_ATTR, {})),
+            [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
+        )
```

Comment:
Can you format all your tests like this ?

```suggestion
    def test_comp_expr(self):
        code = """
            [(lambda x: x**2)(n) for n in range(1, 11)]
        """

        result = eval(cleandoc((code)))
        self.assertEqual(
            result,
            [1, 4, 9, 16, 25, 36, 49, 64, 81, 100],
        )
```

It will gain in clarity to have the code we test on his own line